### DISCUSSION TITLE
Add Profile-Guided Optimization (PGO) support

### DISCUSSION THREAD

**Discussion by zamazan4ik** [SCons/scons#4437] [Ideas]
Hi!

Here I want to discuss the idea of adding to SCons explicit support for building applications with Profile-Guided Optimization (PGO). According to many [articles and benchmarks](https://github.com/zamazan4ik/awesome-pgo), PGO can play a huge role in an application optimization path. However, performing PGO optimization manually via tweaking compiler flags can be error-prone. That's why I think adding such a feature to SCons would be helpful.

Some examples of the current situation in other build systems:

* Cargo: No built-in support but there is awesome [cargo-pgo](https://github.com/Kobzol/cargo-pgo)
* Bazel: Supports ([command-line reference](https://bazel.build/reference/command-line-reference#flag--fdo_instrument))
* CMake: No support yet ([GitLab issue](https://gitlab.kitware.com/cmake/cmake/-/issues/19273))
* Meson: Supports (`b_pgo` in the [docs](https://mesonbuild.com/Builtin-options.html))

**Comment by mwichmann** [comment]
It would be great to have PGO, but I expect it will require considerable thought to implement in a suitably platform-independent way, as I believe all of gcc, clang, msvc and Intel compilers are slightly (or in the case of Microsoft, a lot) different in how they handle this, despite that the general workflow is the same - build-instrumented, training run, rebuild-using-data.  (Thanks for putting it here in Discussions, by the way, that's a good place to track in a longer term way any findings on this).

Just as one example, gcc produces `.gcno` and `.gcda` files, whie msvc produces `.pgc` and `.pgd` files.  SCons would need to be taught about those - how they are made, how they fit into the dependency tree, and so forth.

And what about LTO?

**Reply by zamazan4ik** [reply]
> And what about LTO?

You can use PGO without LTO in all compilers except MSVC. If SCons has no special flag for enabling LTO, it would be great to get it too. But I guess we can create a separate issue for it and track it separately.

**Comment by mwichmann** [comment]
Some interesting bits here, as it shows recipes for building with four different compilers (all on Windows, which means there would still be more to look up).

https://blog.jetbrains.com/clion/2022/05/testing-3-approaches-performance-cpp_apps/
